include/xplat/Starboard/SmartTypes.h

```diff
@@ -305,16 +305,250 @@ bool operator!=(const Any& other, const AutoId<TObj, TLifetimeTraits>& val) {
 
 #ifdef CF_EXPORT // Quick way to detect CoreFoundation.
 namespace woc {
+// Unsafe - a direct non-refcounting store of a refcountable object.
+struct CFLifetimeUnsafe {
+    inline static CFTypeRef store(CFTypeRef* destination, CFTypeRef value) {
+        *destination = value;
+        return value;
+    }
+};
+
+// Retain - a retaining/releasing store of a refcountable object.
+//  - Releases the refcounted value at the destination while never allowing
+//    *destination to hold a non-owning reference.
+struct CFLifetimeRetain {
+    inline static CFTypeRef store(CFTypeRef* destination, CFTypeRef value) {
+        if (*destination == value) {
+            return value;
+        }
+
+        if (*destination) {
+            CFRelease(static_cast<CFTypeRef>(*destination));
+        }
+
+        return *destination = ((value ? CFRetain(value) : nullptr));
+    }
+};
+
+namespace details {
+template <typename T, typename TWrapped>
+class AutoCFRef {
+private:
+    T* ptr;
+
+public:
+    AutoCFRef(T* ptr): ptr(ptr) { }
+
+    operator TWrapped*() const {
+        ptr->attach(nullptr);
+        return &ptr->get();
+    }
+
+    operator void**() const {
```

Comment:
This is nearly 1:1 from the implementation of `WRL::Details::ComPtrRef`, after which this is modelled. Since it's the ref type that's const, we can change the mutable pointer through it (and that is, in fact, what WRL does!)